enhancements/automatic-resource-pruning.md

```diff
@@ -0,0 +1,196 @@
+---
+title: Automatic Resource Pruning
+authors:
+  - "@ryantking"
+reviewers:
+  - "@jmrodri"
+  - "@gallettilance"
+approvers:
+  - "@jmrodri"
+creation-date:2022-01-28
+last-updated: 2022-01-28
+status: implementable
+---
+
+## Summary
+
+This EP aims to provide an easy way for users to operator authors to limit the number of ephemeral resources that may
```

Comment:
awkward, it says "for users to operator authors .." Feels like it should either be "from users to operator authors" or one or the other "for users" or "for operator authors"

Reply:
@fgiloux I think what @ryantking was saying "ephemeral resources" was that their usefulness is short-lived. A job runs, then it's done. It is true that it stays around but it is no long doing anything.  But as this EP is supposed to convey information in an unambiguous way, we should probably look at rephrasing that.

Reply:
@fgiloux I was thinking in terms of how long the resource is "doing work," but I definitely see how that could be vague or inaccurate in this context. Is there a better term that comes to mind for the kinds of resources we're thinking of that the operator is continuously creating?

Reply:
@ryantking @jmrodri sure I did not mean that what was written was wrong. What do you think of "historical"?

Reply:
@fgiloux When I think of "ephemeral" in the context of a Kubernetes resource, I think of something that has a linear progression through some sort of state indicator in the status with one or more final states (usually something like "done" and "error"). Then "historical" data is any of those resources that exist in those final states because no more work is being done on their behalf, they only exist as a record of what work was done.

Reply:
Another consideration is that "historical" data, as you've defined it, may continue to be relevant to an operator after the one-shot execution completes. For example, consider a job that defines a pod that writes some information in its log. 

There are use cases where an operator may use that logged information to derive and manage other on-cluster resources. In that case, that particular job is not ephemeral. If it goes away, the operator will need to re-create and re-run the job so that it can again read its logs and update any downstream state.

In the context of auto-pruning, I think ephemeral is a reasonable choice of words, but be careful conflating specific resource types as being ephemeral or not.

Reply:
@ryantking @joelanford I will try to explain it one last time.
yes ephemeral is about something that can be discarded once it has reached or its use has reached a final state. Think of [ephemeral volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/). The controller discards them as soon as there are released no need for additional logic based on timestamp, number of instances or any other way to decide whether something has to be discarded for not exhausting a limited pool of resources.
yes "historical data" is relevant one-shot execution completes. Think of Tekton [PipelineRuns](https://tekton.dev/docs/pipelines/pipelineruns/) or (VolumeSnapshotContent)https://kubernetes.io/docs/concepts/storage/volume-snapshots/#volume-snapshot-contents
Here what it is about it the later. PipelineRuns consume resources in etcd. When you have 10th of thousands of them you reach a point where etcd and API cache may reach a point where the cluster is not usable anymore. Same for VolumeSnapshotContent with the addition that the storage backend, think Ceph, NFS server, EBS quotas. or resources of any other storage provider may get exhausted first.
Cluster admins want a way to specify that for instance you should not keep more than 10 PipelineRuns per Pipeline or you should remove VolumeSnapshotContent that are older than 30 days.

Reply:
In general, that all seems fine. I'm just saying that the design likely needs to cover the case where an operator needs a particular set of objects to be exempt from pruning. (i.e. an operator author should be able to say, "these GVKs cannot be pruned" or "these GVKS with these label/field selectors cannot be pruned")

That way we can help a cluster admin avoid accidentally configuring a pruning strategy that causes problems for the operator.

Reply:
@ryantking @joelanford @jmrodri After thinking about it a bit more, I think we should use a term other than ephemeral. A resource being ephemeral is not what creates the need for auto-pruning. What creates the need for auto-pruning is the fact that a resource is continuously created and not destroyed. So even though resources targeted for auto-pruning usually are ephemeral (at least as it's defined in this EP), that's not the defining factor. What about using "unbounded resource" with a clear definition in the summary? Something along the lines of an "unbounded resource" is a resource that is continuously created but not destroyed as the operator runs, causing unbounded growth in the number of those resources. I'm not sure "unbounded" is the best term, but I think the definition is more accurate to what properties create the need for auto-pruning.